File: java/kmm/parser/Parser.java
package kmm.parser;

import javafx.util.Pair;
import kmm.utils.Alternative;

import java.util.Optional;
import java.util.function.*;

public class Parser<A> {
    private final Function<String, Optional<Pair<A, String>>> p;

    public Parser(Function<String, Optional<Pair<A, String>>> p){
        this.p = p;
    }

    public Optional<Pair<A, String>> parse(String inp){
        return p.apply(inp);
    }

    // Functor
    public <B> Parser<B> map(Function<? super A, B> f){
        return new Parser<>(inp -> parse(inp).map(pair -> new Pair<>(f.apply(pair.getKey()), pair.getValue())));
    }

    // Applicative
    public static <A> Parser<A> pure(A a){
        return new Parser<>(inp -> Optional.of(new Pair<>(a, inp)));
    }

    public static <A, B> Parser<B> apply(Parser<Function<A, B>> pf, Supplier<Parser<A>> q){
        // q is invoked only if pf parsed successfully.
        return pf.flatMap(f -> q.get().map(f));
    }

    public static <T> Parser<T> apply_u(Parser<UnaryOperator<T>> pf, Supplier<Parser<T>> q){
        // q is invoked only if pf parsed successfully.
        return pf.flatMap(f -> q.get().map(f));
    }

    // Monad
    public <B> Parser<B> flatMap(Function<? super A, Parser<B>> f){
        return new Parser<>(inp -> parse(inp).flatMap(pair -> f.apply(pair.getKey()).parse(pair.getValue())));
    }

    public <B> Parser<B> skip(Parser<B> p){
        return flatMap(a -> p);
    }

    public <B> Parser<B> skip(Supplier<Parser<B>> fp){
        return flatMap(a -> fp.get());
    }

    // Alternative
    public static <A> Parser<A> empty(){
        return new Parser<>(inp -> Optional.empty());
    }

    public Parser<A> orElse(Parser<A> p){
        return new Parser<>(inp -> Alternative.orElseGet(parse(inp), () -> p.parse(inp)));
    }

    public Parser<String> some(){
        return apply(map(c -> s -> c + s), this::many);
    }

    public Parser<String> many(){
        return some().orElse(pure(""));
    }

    public final static Parser<Character> anyChar = new Parser<>(inp -> !inp.isEmpty() ?
        Optional.of(new Pair<>(inp.charAt(0), inp.substring(1))) :
        Optional.empty());

    public static Parser<Character> satisfy(Predicate<Character> f){
        return anyChar.flatMap(c -> f.test(c) ? Parser.pure(c) : Parser.empty());
    }

    public final static Parser<String> spaces = satisfy(Character::isWhitespace).many();

    public static <Open, Close, A> Parser<A> between(Parser<Open> open, Parser<Close>close, Parser<A> p){
        return open.skip(p).flatMap(x -> close.skip(Parser.pure(x)));
    }

    public static <Open, Close, A> Parser<A> between(Parser<Open> open, Parser<Close>close, Supplier<Parser<A>> fp){
        return open.skip(fp).flatMap(x -> close.skip(Parser.pure(x)));
    }

    public Parser<A> token(){
        return between(spaces, spaces, this);
    }

    private static <A> Parser<A> rest(Supplier<Parser<A>> fval, Function<A, Parser<A>> ff, Parser<BinaryOperator<A>> op, A a){
        return op
            .flatMap(f -> fval.get().flatMap(b -> ff.apply(f.apply(a, b))))
            .orElse(Parser.pure(a));
    }

    private Parser<A> rest_l(Parser<BinaryOperator<A>> op, A a){
        return rest(() -> this, b -> rest_l(op, b), op, a);
    }

    private Parser<A> rest_r(Parser<BinaryOperator<A>> op, A a){
        return rest(() ->  chainr1(op), Parser::pure, op, a);
    }

    public Parser<A> chainl1(Parser<BinaryOperator<A>> op){
        return flatMap(a -> rest_l(op, a));
    }

    public Parser<A> chainr1(Parser<BinaryOperator<A>> op){
        return flatMap(a -> rest_r(op, a));
    }
}

File: F#/Parser.fs
module Parser

open System // Char
open Maybe

type Parser<'A>(unp: string -> Option<'A * string>) =
    member this.unp = unp

    member this.Parse (inp: string): Option<'A * string> = this.unp(inp)

    // Functor
    member this.Map (f: 'A -> 'B) = Parser(fun inp -> this.Parse inp |> (</>) (fun (pair: 'A * string) -> (f(fst pair), snd pair)))

    // Monad
    member this.FlatMap (f: 'A -> Parser<'B>): Parser<'B> = Parser(fun inp -> this.Parse inp >>= (fun (pair: 'A * string) -> f(fst pair).Parse(snd pair)))

    member this.Skip (p: Parser<'B>): Parser<'B> = this.FlatMap(fun unit -> p)

    member this.Skip (fp: unit -> Parser<'B>): Parser<'B> = this.FlatMap(fun unit -> fp())

    // Applicative
    static member Pure (x: 'A) = Parser(fun inp -> Some((x, inp)))

    static member Apply (pf: Parser<'A -> 'B>) (q: unit -> Parser<'A>): Parser<'B> = pf.FlatMap(fun f -> q().Map f)

    // Alternative
    static member Empty unit = Parser(fun inp -> None)

    member this.OrElse (p: Parser<'A>): Parser<'A> = Parser(fun inp -> this.Parse inp <|> fun unit -> p.Parse inp)

    static member Optional_s(p: Parser<string>): Parser<string> = p.OrElse(Parser<string>.Pure(""))

    static member Optional_c(p: Parser<char>): Parser<string> = Parser<string>.Optional_s(p.Map(_.ToString()))

    member this.Some unit: Parser<string> =
        ((fun c -> (fun (s: string) -> c.ToString() + s)) |> this.Map |> Parser.Apply) this.Many 

    member this.Many unit: Parser<string> = Parser<string>.Optional_s(this.Some())

    static member anyChar: Parser<char> = Parser(fun inp -> if inp.Length > 0 then Some (inp[0], inp.Substring 1) else None)

    static member Satisfy(pred: char -> bool): Parser<char> = Parser<char>.anyChar.FlatMap(fun (c: char) -> if pred c then Parser.Pure c else Parser<char>.Empty())

    static member spaces: Parser<string> = Parser<char>.Satisfy(Char.IsWhiteSpace).Many()

    static member Between (_open: Parser<'Open>) (_close: Parser<'Close>) (fp: unit -> Parser<'A>): Parser<'A> =
        _open.Skip(fp).FlatMap(fun a -> _close.Skip(Parser.Pure(a)))

    member this.Token unit: Parser<'A> = Parser<'A>.Between Parser<string>.spaces Parser<string>.spaces (fun unit -> this)

    static member Rest (fval: unit -> Parser<'A>) (ff: 'A -> Parser<'A>) (op: Parser<'A -> 'A -> 'A>) (a: 'A): Parser<'A> =
        op.FlatMap(fun f -> fval().FlatMap(fun b -> ff(f a b)))
          .OrElse(Parser.Pure a)

    member this.Rest_l (op: Parser<'A -> 'A -> 'A>) (a: 'A): Parser<'A> = Parser.Rest (fun unit -> this) (this.Rest_l op) op a

    member this.Rest_r (op: Parser<'A -> 'A -> 'A>) = Parser.Rest (fun unit -> this.Chainr1 op) Parser.Pure op

    static member Chainl1 (self: Parser<float>, op: Parser<float -> float -> float>, negate_first: Boolean): Parser<float> =
        self.FlatMap (fun a -> self.Rest_l op (if negate_first then -a else a))

    member this.Chainr1 (op: Parser<'A -> 'A -> 'A>): Parser<'A> = this.FlatMap (this.Rest_r op)
